### PR TITLE
Fix "an orange" not being parsed properly

### PR DESCRIPTION
**Pull request by balvig**

### CODE REVIEW

File: lib/ingreedy/amount_parser.rb [hunk 1]
```diff
@@ -59,7 +59,7 @@ class AmountParser < Parslet::Parser
     private
 
     def word_digits
-      Ingreedy.dictionaries.current.numbers.keys
+      Ingreedy.dictionaries.current.numbers.keys.sort_by(&:length).reverse
```
Comment (davidstosik) [inline lib/ingreedy/amount_parser.rb]: I could be wrong, but I remember seeing `Ingreedy.dictionaries.current.numbers` somewhere else in the gem, maybe it needs the same sorting?

---

Also, I looked for a way to do a reverse sort without having to go through the array twice, and didn't find any method... :( But I found [this StackOverflow answer](https://stackoverflow.com/a/2651028/2341409) that suggests that `.sort_by(&:length).reverse!` might be faster.

```rb
ary = []
10000.times { ary << rand(10000) }
n = 500
Benchmark.bm(20) do |x|
  x.report("sort.reverse")       { n.times { ary.sort.reverse } }
  x.report("sort.reverse!")   { n.times { ary.sort.reverse! } }
end
```

Comment (davidstosik) [inline lib/ingreedy/amount_parser.rb]: @balvig 
~I could be wrong, but I remember seeing `Ingreedy.dictionaries.current.numbers` somewhere else in the gem, maybe it needs the same sorting?~ *I was wrong, the other place does not need sorting.*

---

Also, I looked for a way to do a reverse sort without having to go through the array twice, and didn't find any method... :( But I found [this StackOverflow answer](https://stackoverflow.com/a/2651028/2341409) that suggests that `.sort_by(&:length).reverse!` might be faster.

```rb
ary = []
10000.times { ary << rand(10000) }
n = 500
Benchmark.bm(20) do |x|
  x.report("sort.reverse") { n.times { ary.sort.reverse } }
  x.report("sort.reverse!") { n.times { ary.sort.reverse! } }
end

#                            user     system      total        real
# sort.reverse           1.000000   0.150000   1.150000 (  1.253906)
# sort.reverse!          0.610000   0.010000   0.620000 (  0.625165)
```

---

Maybe we could memoize the result too?

Comment (davidstosik) [reply lib/ingreedy/amount_parser.rb]: Maybe we could also sort `numbers` when the current `Ingreedy::Dictionary` is [initialized](https://github.com/cookpad/ingreedy/blob/master/lib/ingreedy/dictionary.rb#L8)?

Comment (balvig) [reply lib/ingreedy/amount_parser.rb]: @davidstosik ah yes, I think I wanted to stay away from micro-optimizing as long as the performance specs in global-web were passing. Was following the [pattern here](https://github.com/cookpad/ingreedy/blob/master/lib/ingreedy/root_parser.rb#L135-L140).

Sorting upon initialization sounds like a good idea though, let me take a look!

**Comment by balvig** [comment]
@davidstosik ♻️  please (not sure why Travis is [choking on 2.0](https://travis-ci.org/cookpad/ingreedy/builds/259639646?utm_source=github_status&utm_medium=notification) even though I removed it)

File: lib/ingreedy/dictionary.rb [hunk 2]
```diff
@@ -3,12 +3,12 @@ class Dictionary
     attr_reader :units, :numbers, :prepositions, :range_separators
     attr_reader :imprecise_amounts
 
-    def initialize(entries = {})
-      @units = entries[:units] || raise("No units found in dictionary")
-      @numbers = entries[:numbers] || {}
-      @prepositions = entries[:prepositions] || []
-      @range_separators = entries[:range_separators] || %w{- ~}
-      @imprecise_amounts = entries[:imprecise_amounts] || []
+    def initialize(units:, numbers: {}, prepositions: [], range_separators: %w{- ~}, imprecise_amounts: [])
```
Comment (davidstosik) [inline lib/ingreedy/dictionary.rb:6]: 👍

File: lib/ingreedy/dictionary.rb [hunk 3]
```diff
@@ -34,5 +34,11 @@ def vulgar_fractions
         "\u215E" => "7/8",
       }
     end
+
+    private
+
+      def sort_by_length(hash)
+        hash.sort_by { |key, val| -key.length }.to_h
```
Comment (davidstosik) [inline lib/ingreedy/dictionary.rb:41]: Optimum! 👍

**Review by davidstosik — APPROVED** [review]
Looks great!

**Comment by davidstosik** [comment]
@balvig Maybe Travis uses `master`'s version of `.travis.yml` to initialize?

**Pull request merged**